zebra/zserv.c

```diff
@@ -483,36 +516,35 @@ static void zserv_client_event(struct zserv *client,
  * as the task argument.
  *
  * Each message is popped off the client's input queue and the action associated
- * with the message is executed. This proceeds until there are no more messages,
- * an error occurs, or the processing limit is reached.
+ * with the message is executed. This proceeds until an error occurs, or the
+ * processing limit is reached.
  *
  * The client's I/O thread can push at most zrouter.packets_to_process messages
  * onto the input buffer before notifying us there are packets to read. As long
  * as we always process zrouter.packets_to_process messages here, then we can
  * rely on the read thread to handle queuing this task enough times to process
  * everything on the input queue.
+ *
+ * If the client ibuf has maximum items(packets_to_process) then the main thread
+ * schedules a call to the client pthread to pump in more items on the client's
+ * input buffer fifo.
  */
 static void zserv_process_messages(struct event *thread)
 {
 	struct zserv *client = EVENT_ARG(thread);
 	struct stream *msg;
 	struct stream_fifo *cache = stream_fifo_new();
 	uint32_t p2p = zrouter.packets_to_process;
-	bool need_resched = false;
+	size_t client_ibuf_fifo_cnt;
 
 	frr_with_mutex (&client->ibuf_mtx) {
+		client_ibuf_fifo_cnt = stream_fifo_count_safe(client->ibuf_fifo);
 		uint32_t i;
 		for (i = 0; i < p2p && stream_fifo_head(client->ibuf_fifo);
 		     ++i) {
 			msg = stream_fifo_pop(client->ibuf_fifo);
 			stream_fifo_push(cache, msg);
 		}
-
-		/* Need to reschedule processing work if there are still
```

Comment:
not really happy to see this lost - I think it risks making this system brittle.
I'd prefer that we make a distinction between the fifo management, limiting the amount of buffering we'll do in this path, and the work or event management, limiting the amount of work we'll do in a single processing callback.

Reply:
Incorporated comments as discussed. Summary of changes available in the commit message..